### ISSUE TITLE
Bug: Qwen3 native-KV decode cost scales with the 40960-row default KV capacity, not sequence length

### ISSUE THREAD

**Issue by LiRunGuo** [NVIDIA/TensorRT-Model-Connect#1409]
### Affected area

Inference output or task execution

### TensorRT-Model-Connect version

`main` at `393ab02f`

### Installation method

Development container

### Model and revision

`Qwen/Qwen3-0.6B` (Hub `main` snapshot downloaded 2026-09-21) and a local `Qwen3-4B` checkpoint (revision not recorded); both `fp16`.

### Environment

NVIDIA H200 (SM90), driver 580.159.03, Ubuntu 24.04 container built from `Dockerfile.dev.x86` (TensorRT 11.1.0.106, CUDA 13.3, Python 3.12). Timings were taken on an otherwise idle GPU.

### Minimal reproducer

```bash
python -m tensorrt_model_connect build Qwen/Qwen3-0.6B --precision fp16 --output default.bundle
python -m tensorrt_model_connect build Qwen/Qwen3-0.6B --precision fp16 \
  --max-sequence-length 2048 --output s2048.bundle
for bundle in default.bundle s2048.bundle; do
  trtmc run "$bundle" --runtime-root "$TRTMC_BUILD_DIR" \
    --prompt "Explain what TensorRT does in one sentence." --max-new-tokens 128
done
# Compare decode_ms / (len(token_ids) - 1) in the JSON output.
```

### Observed behavior

With a default build, native Qwen3 decode latency is proportional to the bundle's KV-cache **capacity**, not to the tokens actually in the cache. The default capacity is the checkpoint's full `max_position_embeddings` (40960 for Qwen3), so short and medium requests pay for the full context window on every decode step.

On an H200, a 10-token prompt generating 128 greedy tokens decodes **4.5x slower** for Qwen3-0.6B and **2.7x slower** for Qwen3-4B than the same checkpoint built with a smaller `--max-sequence-length`. Qwen3-4B token IDs are identical at both capacities; Qwen3-0.6B differs only at an fp16 near-tie, detailed below. This may contribute to the gaps tracked in #901 and the other `[Performance]` issues, although those were measured on Thor/GB300 with 10-token outputs.

NVIDIA H200 (SM90), driver 580.159.03, `Dockerfile.dev.x86` image (TensorRT 11.1.0.106, CUDA 13.3), repository head `393ab02f`, `--precision fp16`, split prefill/decode layout. Prompt `"Explain what TensorRT does in one sentence."`, `trtmc run ... --max-new-tokens 128`, decode ms divided by 127 decode steps. Each value is the median of 5 interleaved rounds on an otherwise idle GPU (min/max within ±1%).

| Model | `--max-sequence-length` | decode ms/token | prefill ms |
| --- | --- | --- | --- |
| Qwen3-0.6B | 2048 | 1.73 | 27.6 |
| Qwen3-0.6B | 8192 | 2.76 | 27.6 |
| Qwen3-0.6B | 16384 | 4.29 | 31.3 |
| Qwen3-0.6B | default (40960) | **7.76** | 40.1 |
| Qwen3-4B | 4096 | 4.35 | 40.4 |
| Qwen3-4B | default (40960) | **11.67** | 64.2 |

Decode time grows linearly with capacity, about 0.15 ms per 1K rows for Qwen3-0.6B. Qwen3-4B produced identical token IDs at both capacities. Qwen3-0.6B's 2048/16384 builds diverge from the others at generated token 30, where the fp32 Transformers reference's top-2 logits differ by only 0.023 (19.2698 vs 19.2467), so that divergence is an fp16 near-tie rather than a capacity effect. The default-capacity bundle matched Transformers fp32 greedy output for all 128 tokens.

**Cause**

- `families/qwen/model.py:80-83`: `default_max_cache_length` returns `config.max_position_embeddings` for native-KV-eligible Qwen3, and `build()` uses it when `--max-sequence-length` is omitted (`model.py:271`).
- `families/qwen/native_kv_attention_builder.py`: `add_active_prefix_causal_masks` builds masks over `cache_capacity` key positions (line 95), and `add_explicit_masked_grouped_query_attention` (line 158) computes the score matmul, mask select, softmax and context matmul over the whole static `[1, Hkv, capacity, D]` cache. It also sanitizes the entire K and V caches with `select` every step (lines 211 and 216). The active length only enters as a mask, so the work per step is O(capacity).

`families/llama/model.py:188` and `families/smollm3/model.py:148` use the same `max_position_embeddings` default with a duplicated native-KV attention builder. For Llama 3.1 that default is 131072 rows. I have not measured those families yet.

**A small, safe part: the K sanitizing select is redundant**

The score mask already replaces every position that is not both causal and active with the blocked value, and the K `select` only zeros positions that are inactive. So the K select cannot change the softmax input, including when an unwritten row holds NaN or Inf. Only the V `select` protects the context matmul from `0 * NaN`. In an out-of-tree build that fed `key_grouped` directly into the score matmul, Qwen3-0.6B at default capacity decoded 128 tokens with identical token IDs:

| Variant (Qwen3-0.6B, capacity 40960) | decode ms/token | prefill ms |
| --- | --- | --- |
| current | 7.76 | 40.1 |
| without the K select | 6.90 (-11%) | 38.0 |
| without K and V selects (timing only; not safe without a zeroed cache) | 7.13 | n/a |

Removing the V select as well does not help further, so the selects are not the dominant cost. The O(capacity) attention is.

### Expected behavior

Per-token decode cost should follow the number of cached tokens rather than the reserved capacity, or at least the capacity/latency trade-off of the default build should be explicit to users.

Possible directions (for maintainers to judge):

1. Document `--max-sequence-length` as a latency knob. The README and several guides pass it in examples (for instance `16384` in the README), but `cli-reference.md` only calls it an "optional family-consumed sequence bound". Nothing says that for native-KV text families it also sets the per-token decode cost, and builds that omit it get the full context window.
2. Drop the redundant K `select` (family-local, bit-identical in the run above, about 11% here).
3. Bound decode attention by the active length while keeping static shapes for CUDA graphs. For example, a small set of attention-window buckets as optimization profiles of one decode engine (weights shared, as in the dual-profile layout), each statically slicing the full-capacity cache, with the runtime selecting the smallest bucket that covers the current position.
4. Or reconsider the default capacity, which is a behavior change and a product decision.

I am happy to send the documentation or K-select change as a family-local PR if that direction is wanted, and to measure Llama 3.1 8B at its 131072-row default.

### Relevant logs

```text
[trtmc] KV cache rows=40960 (bundle max=40960)
[trtmc.engine_timing] label="qwen decoder" execute_ms=960.129696 launches=128   # default capacity
[trtmc.engine_timing] label="qwen decoder" execute_ms=181.797151 launches=128   # --max-sequence-length 2048
```

### Submission checks

- [x] I searched open and closed issues and found no duplicate.
- [x] This is not a security vulnerability.
- [x] I removed secrets, credentials, private URLs, and restricted artifacts.

**Comment by LiRunGuo** [comment]
Follow-up with the Llama measurement I offered. `meta-llama/Llama-3.1-8B-Instruct` (local snapshot), `--precision bf16` (the native-KV route), same H200 environment and prompt, 128 greedy tokens, median of 5 interleaved rounds on an idle GPU (min/max within ±0.4%):

| `--max-sequence-length` | decode ms/token | prefill ms |
| --- | --- | --- |
| 4096 | 11.26 | 79.3 |
| default (131072) | **26.94** | 97.8 |

Token IDs are identical. The default capacity costs 2.4x per decode step here.

Building with the defaults does not reach this path at all. Without `--precision`, Llama builds in fp32 and the build fails; with `--precision fp16`, the bundle builds but cannot run on a 141 GB GPU. I have filed that separately as #1414 because it is a different defect.
